lib/vtls/rustls.c

```diff
@@ -0,0 +1,409 @@
+  uint8_t tlsbuf[16384];
+  ssize_t n = 0;
+  ssize_t tls_bytes_read = 0;
+  ssize_t tls_bytes_processed = 0;
+  ssize_t plain_bytes_copied = 0;
+  int rustls_result = 0;
+
+  bzero(tlsbuf, sizeof(tlsbuf));
+  tls_bytes_read = read(sockfd, tlsbuf, sizeof(tlsbuf));
+  if(tls_bytes_read == 0) {
+    failf(data, "rustls_recv: EOF reading from socket");
+    *err = CURLE_READ_ERROR;
+    return -1;
+  }
+  else if(tls_bytes_read < 0) {
+    if(SOCKERRNO == EAGAIN || SOCKERRNO == EWOULDBLOCK) {
+      infof(data, "rustls_recv: EAGAIN or EWOULDBLOCK\n");
+      *err = CURLE_AGAIN;
+      return 0;
+    }
+    failf(data, "rustls_recv: reading from socket: %s", strerror(SOCKERRNO));
+    *err = CURLE_READ_ERROR;
+    return -1;
+  }
+  infof(data, "rustls_recv: read %d bytes from socket\n", tls_bytes_read);
+
+  /*
+  * Now pull those bytes from the buffer into ClientSession.
+  */
+  while(tls_bytes_processed < tls_bytes_read) {
+    n = rustls_client_session_read_tls(session,
+      (uint8_t *)tlsbuf + tls_bytes_processed,
+      tls_bytes_read - tls_bytes_processed);
+    if(n == 0) {
+      infof(data, "rustls_recv: EOF from rustls_client_session_read_tls\n");
+      break;
+    }
+    else if(n < 0) {
+      failf(data, "rustls_recv: error in rustls_client_session_read_tls");
+      *err = CURLE_READ_ERROR;
+      return -1;
+    }
+
+    rustls_result = rustls_client_session_process_new_packets(session);
+    if(rustls_result != RUSTLS_RESULT_OK) {
+      failf(data, "rustls_recv: error in process_new_packets");
+      *err = CURLE_READ_ERROR;
+      return -1;
+    }
+
+    tls_bytes_processed += n;
+    backend->data_pending = TRUE;
+  }
+
+  while((size_t)plain_bytes_copied < plainlen) {
+    n = rustls_client_session_read(session,
+      (uint8_t *)plainbuf + plain_bytes_copied,
+      plainlen - plain_bytes_copied);
+    if(n == 0) {
+      /* rustls returns 0 from client_session_read to mean "all currently
+        available data has been read." If we bring in more ciphertext with
+        read_tls, more plaintext will become available. So don't tell curl
+        this is an EOF. Instead, say "come back later." */
+      infof(data, "rustls_recv: EOF from rustls_client_session_read\n");
+      backend->data_pending = FALSE;
+      break;
+    }
+    else if(n < 0) {
+      failf(data, "rustls_recv: error in rustls_client_session_read");
+      *err = CURLE_READ_ERROR;
+      return -1;
+    }
+    else {
+      plain_bytes_copied += n;
+    }
+  }
+
+  /* If we wrote out 0 plaintext bytes, it might just mean we haven't yet
+     read a full TLS record. Return CURLE_AGAIN so curl doesn't treat this
+     as EOF. */
+  if(plain_bytes_copied == 0) {
+    *err = CURLE_AGAIN;
+    return -1;
+  }
+
+  return plain_bytes_copied;
+}
+
+/*
+ * Write n bytes from buf to the provided fd, retrying short writes until
+ * we finish or hit an error. Assumes fd is blocking and therefore doesn't
+ * handle EAGAIN. Returns 0 for success or 1 for error.
+ */
```

Comment:
This sounds suspicious. libcurl should *always* use non-blocking sockets and this kind of loop is not a good sign.